cmd/natc/ippool/consensusippool.go

```diff
@@ -0,0 +1,343 @@
+// the traffic, but writes must be sent to the consensus peers.
+//
+// To handle expiry we write on reads, to update the last-used-date, but we do that after we've returned a response.
+type ConsensusIPPool struct {
+	IPSet      *netipx.IPSet
+	perPeerMap *syncs.Map[tailcfg.NodeID, *consensusPerPeerState]
+	consensus  commandExecutor
+}
+
+func NewConsensusIPPool(ipSet *netipx.IPSet) *ConsensusIPPool {
+	return &ConsensusIPPool{
+		IPSet:      ipSet,
+		perPeerMap: &syncs.Map[tailcfg.NodeID, *consensusPerPeerState]{},
+	}
+}
+
+// DomainForIP is part of the IPPool interface. It returns a domain for a given IP address, if we have
+// previously assigned the IP address to a domain for the node that is asking. Otherwise it logs and returns the empty string.
+func (ipp *ConsensusIPPool) DomainForIP(from tailcfg.NodeID, addr netip.Addr, updatedAt time.Time) (string, bool) {
+	ww, ok := ipp.retryDomainLookup(from, addr, 0)
+	if !ok {
+		return "", false
+	}
+	go func() {
+		err := ipp.markLastUsed(from, addr, ww.Domain, updatedAt)
+		if err != nil {
+			panic(err)
+		}
+	}()
+	return ww.Domain, true
+}
+
+// retryDomainLookup tries to lookup the domain for this IP+node. If it can't find the node or the IP it
+// tries again up to 5 times, with exponential backoff.
+// The raft lib will tell the leader that a log entry has been applied to a quorum of nodes, sometimes before the
+// log entry has been applied to the local state. This means that in our case the traffic on an IP can arrive before
+// we have the domain for which that IP applies stored.
+func (ipp *ConsensusIPPool) retryDomainLookup(from tailcfg.NodeID, addr netip.Addr, n int) (whereWhen, bool) {
+	ps, foundPeerState := ipp.perPeerMap.Load(from)
+	if foundPeerState {
+		ps.mu.Lock()
+		ww, foundDomain := ps.addrToDomain.Lookup(addr)
+		ps.mu.Unlock()
+		if foundDomain {
+			return ww, true
+		}
+	}
+	if n > 4 {
+		if !foundPeerState {
+			log.Printf("DomainForIP: peer state absent for: %d", from)
+		} else {
+			log.Printf("DomainForIP: peer state doesn't recognize addr: %s", addr)
+		}
+		return whereWhen{}, false
+	}
+	timeToWait := 100
+	for i := 0; i < n; i++ {
+		timeToWait *= 2
+	}
+	time.Sleep(time.Millisecond * time.Duration(timeToWait))
+	return ipp.retryDomainLookup(from, addr, n+1)
+}
+
+// StartConsensus is part of the IPPool interface. It starts the raft background routines that handle consensus.
+func (ipp *ConsensusIPPool) StartConsensus(ctx context.Context, ts *tsnet.Server, clusterTag string) error {
+	cfg := tsconsensus.DefaultConfig()
+	cfg.ServeDebugMonitor = true
+	cns, err := tsconsensus.Start(ctx, ts, ipp, clusterTag, cfg)
+	if err != nil {
+		return err
+	}
+	ipp.consensus = cns
+	return nil
+}
+
+type whereWhen struct {
+	Domain   string
+	LastUsed time.Time
+}
+
+type consensusPerPeerState struct {
+	domainToAddr map[string]netip.Addr
+	addrToDomain *bart.Table[whereWhen]
```

Comment:
fwiw, if it ends up being simpler, I'd consider just putting this in a map of netip.Addr -> whereWhen, we're not gaining much from bart here, particularly as the path to this table always produces commands.